src/js/stores/document.js

```diff
@@ -166,18 +173,37 @@ define(function (require, exports, module) {
             if (suppressChange) {
                 return;
             }
-
-            // If some selected layer remains uninitialized, a new document will
-            // come along shortly. Wait until then to trigger the change event.
-            var initialized = nextDocument.layers && nextDocument.layers.selected
-                .every(function (layer) {
-                    return layer.initialized;
-                });
-
-            if (initialized) {
-                // Include the document in the event payload
-                this.emit(changeEventName || "change", nextDocument);
+            
+            var promise = Promise.bind(this);
+
+            // Notify change of layer selection state.
+            if (oldDocument) {
+                promise = promise.then(function () {
+                    oldDocument.layers.layers.forEach(function (layer) {
+                        var nextLayer = nextDocument.layers.byID(layer.id),
+                            callback = this._layerIDToStateListenner[layer.id];
+                        
+                        if (nextLayer && callback &&
+                            (layer.selected !== nextLayer.selected || layer.expanded !== nextLayer.expanded)) {
+                            callback(nextLayer.selected, nextLayer.expanded);
+                        }
+                    }, this);
+                }).delay(0); // Performance Hack: delay emitting the `change` event to promote browser paint event.
             }
+            
+            promise.then(function () {
```

Comment:
Just to be clear, the above `delay(0)` call will not apply to the then-handler below, right?